File: internal/knowledge/retrieval_scoring.go
package knowledge

import (
	"math"
	"strings"
	"unicode"

	"golang.org/x/text/unicode/norm"
)

const (
	bm25K1              = 1.5
	bm25B               = 0.75
	patternsFieldWeight = 4.0
	titleFieldWeight    = 3.0
	contentFieldWeight  = 1.0
)

type retrievalBM25Index struct {
	patterns bm25FieldIndex
	titles   bm25FieldIndex
	contents bm25FieldIndex
}

type bm25FieldIndex struct {
	documents []bm25Document
	idf       map[string]float64
	avgLength float64
}

type bm25Document struct {
	termFrequency map[string]int
	length        int
}

func newRetrievalBM25Index(chunks []KBChunk) retrievalBM25Index {
	patterns := make([]string, 0, len(chunks))
	titles := make([]string, 0, len(chunks))
	contents := make([]string, 0, len(chunks))
	for _, chunk := range chunks {
		patterns = append(patterns, strings.Join(chunk.QuestionPatterns, " "))
		titles = append(titles, chunk.Title)
		contents = append(contents, chunk.Content)
	}
	return retrievalBM25Index{
		patterns: newBM25FieldIndex(patterns),
		titles:   newBM25FieldIndex(titles),
		contents: newBM25FieldIndex(contents),
	}
}

func newBM25FieldIndex(values []string) bm25FieldIndex {
	documents := make([]bm25Document, 0, len(values))
	documentFrequency := map[string]int{}
	totalLength := 0
	for _, value := range values {
		tokens := tokenizeRetrievalText(value)
		document := bm25Document{termFrequency: map[string]int{}, length: len(tokens)}
		seen := map[string]struct{}{}
		for _, token := range tokens {
			document.termFrequency[token]++
			seen[token] = struct{}{}
		}
		for token := range seen {
			documentFrequency[token]++
		}
		totalLength += document.length
		documents = append(documents, document)
	}
	avgLength := 1.0
	if len(documents) > 0 && totalLength > 0 {
		avgLength = float64(totalLength) / float64(len(documents))
	}
	idf := map[string]float64{}
	n := float64(len(documents))
	for token, df := range documentFrequency {
		idf[token] = math.Log((n-float64(df)+0.5)/(float64(df)+0.5) + 1)
	}
	return bm25FieldIndex{documents: documents, idf: idf, avgLength: avgLength}
}

func (field bm25FieldIndex) score(chunkIndex int, queryTokens []string) float64 {
	if chunkIndex < 0 || chunkIndex >= len(field.documents) {
		return 0
	}
	document := field.documents[chunkIndex]
	if document.length == 0 {
		return 0
	}
	score := 0.0
	for _, token := range queryTokens {
		tf := document.termFrequency[token]
		if tf == 0 {
			continue
		}
		idf := field.idf[token]
		tfFloat := float64(tf)
		denominator := tfFloat + bm25K1*(1-bm25B+bm25B*float64(document.length)/field.avgLength)
		if denominator == 0 {
			continue
		}
		score += idf * (tfFloat * (bm25K1 + 1)) / denominator
	}
	return score
}

func tokenizeRetrievalText(value string) []string {
	normalized := NormalizeQuery(value)
	if normalized == "" {
		return nil
	}
	var tokens []string
	for _, segment := range strings.Fields(normalized) {
		if segment == "" {
			continue
		}
		if isASCIIAlnumSegment(segment) {
			tokens = append(tokens, segment)
			continue
		}
		runes := []rune(segment)
		for n := 2; n <= 3; n++ {
			if len(runes) < n {
				continue
			}
			for index := 0; index <= len(runes)-n; index++ {
				tokens = append(tokens, string(runes[index:index+n]))
			}
		}
	}
	return tokens
}

// NormalizeQuery is shared by the runtime retriever and eval parity tests; keep
// its preprocessing semantics aligned with the BM25 scorer in this file.
func NormalizeQuery(value string) string {
	value = strings.ToLower(norm.NFKC.String(value))
	var builder strings.Builder
	lastWasSpace := true
	for _, r := range value {
		switch {
		case isASCIIAlnum(r) || isCJK(r):
			builder.WriteRune(r)
			lastWasSpace = false
		case unicode.IsSpace(r):
			if !lastWasSpace {
				builder.WriteByte(' ')
				lastWasSpace = true
			}
		default:
			// Punctuation and symbols are intentionally dropped.
		}
	}
	return strings.TrimSpace(builder.String())
}

func isASCIIAlnumSegment(value string) bool {
	for _, r := range value {
		if !isASCIIAlnum(r) {
			return false
		}
	}
	return value != ""
}

func isASCIIAlnum(r rune) bool {
	return (r >= 'a' && r <= 'z') || (r >= '0' && r <= '9')
}

func isCJK(r rune) bool {
	for _, rng := range cjkRanges {
		if r >= rng.lo && r <= rng.hi {
			return true
		}
	}
	return false
}

var cjkRanges = []struct {
	lo rune
	hi rune
}{
	{0x3400, 0x4DBF},
	{0x4E00, 0x9FFF},
	{0xF900, 0xFAFF},
	{0x20000, 0x2A6DF},
	{0x2A700, 0x2EBEF},
	{0x30000, 0x3134F},
}

File: scripts/rag_w0/retrieval_scoring.py
#!/usr/bin/env python3
"""Shared W0 retrieval scoring primitives."""

from __future__ import annotations

from collections import Counter
from dataclasses import dataclass
import math
import re
import unicodedata
from typing import Any


BM25_K1 = 1.5
BM25_B = 0.75
DEFAULT_THRESHOLD = 0.5
PATTERNS_WEIGHT = 4.0
TITLE_WEIGHT = 3.0
CONTENT_WEIGHT = 1.0
AREA_BONUS = 2.0


@dataclass(frozen=True)
class BM25Document:
    term_frequency: Counter[str]
    length: int


@dataclass(frozen=True)
class BM25FieldIndex:
    documents: list[BM25Document]
    idf: dict[str, float]
    avg_length: float

    def score(self, chunk_index: int, query_tokens: list[str]) -> float:
        if chunk_index < 0 or chunk_index >= len(self.documents):
            return 0.0
        document = self.documents[chunk_index]
        if document.length == 0:
            return 0.0
        score = 0.0
        for token in query_tokens:
            tf = document.term_frequency.get(token, 0)
            if tf == 0:
                continue
            denominator = tf + BM25_K1 * (1 - BM25_B + BM25_B * document.length / self.avg_length)
            if denominator == 0:
                continue
            score += self.idf.get(token, 0.0) * (tf * (BM25_K1 + 1)) / denominator
        return score


class BM25Index:
    def __init__(self, chunks: list[dict[str, Any]]) -> None:
        self.patterns = _field_index([" ".join(str(item) for item in chunk.get("question_patterns") or []) for chunk in chunks])
        self.titles = _field_index([str(chunk.get("title") or "") for chunk in chunks])
        self.contents = _field_index([str(chunk.get("content") or "") for chunk in chunks])

    def score_chunk(self, *, query_tokens: list[str], product_area: str, chunk_index: int, chunk: dict[str, Any]) -> float:
        if not query_tokens:
            return 0.0
        score = (
            PATTERNS_WEIGHT * self.patterns.score(chunk_index, query_tokens)
            + TITLE_WEIGHT * self.titles.score(chunk_index, query_tokens)
            + CONTENT_WEIGHT * self.contents.score(chunk_index, query_tokens)
        )
        if score <= 0:
            return 0.0
        if product_area and product_area == str(chunk.get("product_area") or "").strip().lower():
            score += AREA_BONUS
        return score


def normalize_text(value: str) -> str:
    normalized = unicodedata.normalize("NFKC", value).lower()
    chars: list[str] = []
    last_was_space = True
    for char in normalized:
        if char.isascii() and char.isalnum():
            chars.append(char)
            last_was_space = False
        elif _is_cjk(char):
            chars.append(char)
            last_was_space = False
        elif char.isspace():
            if not last_was_space:
                chars.append(" ")
                last_was_space = True
        else:
            continue
    return "".join(chars).strip()


def tokenize_text(value: str) -> list[str]:
    normalized = normalize_text(value)
    if not normalized:
        return []
    tokens: list[str] = []
    for segment in normalized.split():
        if _ASCII_SEGMENT_RE.fullmatch(segment):
            tokens.append(segment)
            continue
        chars = list(segment)
        for size in (2, 3):
            if len(chars) < size:
                continue
            for index in range(0, len(chars) - size + 1):
                tokens.append("".join(chars[index : index + size]))
    return tokens


def _field_index(values: list[str]) -> BM25FieldIndex:
    documents: list[BM25Document] = []
    document_frequency: Counter[str] = Counter()
    total_length = 0
    for value in values:
        tokens = tokenize_text(value)
        term_frequency = Counter(tokens)
        documents.append(BM25Document(term_frequency=term_frequency, length=len(tokens)))
        document_frequency.update(term_frequency.keys())
        total_length += len(tokens)
    avg_length = (total_length / len(documents)) if documents and total_length else 1.0
    count = float(len(documents))
    idf = {
        token: math.log((count - df + 0.5) / (df + 0.5) + 1)
        for token, df in document_frequency.items()
    }
    return BM25FieldIndex(documents=documents, idf=idf, avg_length=avg_length)


def _is_cjk(char: str) -> bool:
    codepoint = ord(char)
    return any(start <= codepoint <= end for start, end in _CJK_RANGES)


_ASCII_SEGMENT_RE = re.compile(r"[a-z0-9]+")
_CJK_RANGES = (
    (0x3400, 0x4DBF),
    (0x4E00, 0x9FFF),
    (0xF900, 0xFAFF),
    (0x20000, 0x2A6DF),
    (0x2A700, 0x2EBEF),
    (0x30000, 0x3134F),
)
